pkg/build/kube/tarbits.go

```diff
@@ -0,0 +1,425 @@
+// a local file path, using the prefix "file://" explicitly indicates the given
+// tarRoot is a local file path.
+//
+// The given tarRoot may also take the format tarRoot::localPath where the
+// tarRoot is separated from the local path where the remote files will be
+// downloaded.
+//
+// If no explicit local path is provided then remote files are downloaded to
+// "$HOME/.kind/tarbits/VERSION/GOOS/GOARCH" where VERSION is the value
+// extracted from the "version" file in the "kubernetes.tar.gz" tarball.
+//
+// If a file is already present on disk with the same size as the remote
+// equivalent, the file will not be downloaded again.
+func NewTarBits(tarRoot string) (bits Bits, err error) {
+	tarBits := &TarBits{}
+
+	if strings.HasPrefix(tarRoot, "file://") {
+		tarBits.tarRoot = strings.Replace(tarRoot, "file://", "", 1)
+		tarBits.tbitsType = tarBitsLocal
+	} else if strings.HasPrefix(tarRoot, "http:") ||
+		strings.HasPrefix(tarRoot, "https:") {
+		tarBits.remoteURI = tarRoot
+		tarBits.tbitsType = tarBitsHTTP
+	} else if ok, _ := regexp.MatchString(
+		`^(v\d+\.\d+\.\d+[\w.\-+]*)(::.+)?$`, tarRoot); ok {
+		tarBits.remoteURI = fmt.Sprintf("%s/release/%s", releaseURIPrefix, tarRoot)
+		tarBits.tbitsType = tarBitsSemVer
+	} else if strings.HasPrefix(tarRoot, "ci/") {
+		tarBits.remoteURI = fmt.Sprintf("%s/%s", ciURIPrefix, tarRoot)
+		tarBits.tbitsType = tarBitsCI
+	} else if strings.HasPrefix(tarRoot, "release/") {
+		tarBits.remoteURI = fmt.Sprintf("%s/%s", releaseURIPrefix, tarRoot)
+		tarBits.tbitsType = tarBitsRelease
+	} else {
+		tarBits.tarRoot = tarRoot
+		tarBits.tbitsType = tarBitsLocal
+	}
+
+	// Check to see if there's a local path encoded into the remote URI.
+	if tarBits.tarRoot == "" {
+		parts := strings.Split(tarBits.remoteURI, "::")
+		tarBits.remoteURI = parts[0]
+		if len(parts) > 1 {
+			tarBits.tarRoot = parts[1]
+		}
+	}
+
+	return tarBits, nil
+}
+
+// Build implements Bits.Build
+func (b *TarBits) Build() error {
+	// If this is a CI or Release URI then check to see if the URI is just a
+	// pointer to the actual version.
+	if b.tbitsType == tarBitsCI || b.tbitsType == tarBitsRelease {
+		if err := b.getRealURIForCIOrRelease(); err != nil {
+			return err
+		}
+	}
+
+	// If the files are not local then download them.
+	if b.tbitsType != tarBitsLocal {
+
+		// Allow the OS and architecture to be configurable based on the
+		// GOOS and GOARCH environment variables rather than the GOOS and
+		// GOARCH of the running process.
+		goos := os.Getenv("GOOS")
+		goarch := os.Getenv("GOARCH")
+		if goos == "" {
+			goos = util.GetOS()
+		}
+		if goarch == "" {
+			goarch = util.GetArch()
+		}
+
+		// Get the version from the remote "kubernetes.tar.gz"
+		version, err := b.getVersion()
+		if err != nil {
+			return err
+		}
+		fmt.Printf("Building version %s\n", version)
+
+		// Get the location where to the remote files are cached locally.
+		if b.tarRoot == "" {
+			currentUser, err := user.Current()
```

Comment:
this is entering a retrospective minefield of the standard go library.
nowadays this should be portable to all the OSes..

Reply:
Hi @neolit123,

I'm not sure if your comment was more opining the portability or if you're suggesting a change?

FWIW the `user.Current()` call *is* [portable](https://golang.org/src/os/user/). I had originally used [`os.UserHomeDir`](https://golang.org/pkg/os/#UserHomeDir), but the CI builds failed. I don't know when `UserHomeDir` was introduced, but it must be in a version later than the one used to build Kind.

Reply:
it was just a comment.
the go in the CI that builds kind *should be* 1.12.1.

alternatively this can be just a mandatory folder specified by the user or defaulting to `./` instead of defaulting to the user home path.

Reply:
Hi @neolit123,

I opted to avoid adding any new flags since @BenTheElder indicated he would like all the flags to apply to all the commands.

I'm fervently against the idea of `./` because of the size of what's being downloaded. It's much friendlier to local development to support a well-defined path for the tarbits to avoid re-downloading files when they're already cached locally.

I really don't think it's a problem to use the user's home directory. It's a fairly standard approach for storing app data.

Reply:
Why should they be cached at all? When do entries get evicted?

Reply:
Hi @BenTheElder,

See my comment above. I agree that perhaps caching `ci/` builds is unnecessary. Caching release builds, however, seems to be not the worst idea. At the very minimum having an opt-in mode for caching would be nice for local devs.